### PR TITLE
feat: add useBodySize hook

### PR DESCRIPTION
**Pull request by carbongo**
# Description

A new sensor hook _useBodySize_ added. It has various use-cases, e.g. when a whole page width and/or height are needed within Components. Personally I used it for a "smart" scroll-up button, which has a full stop on a certain point of the page.

## Type of change

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] **Breaking change** _(fix or feature that would cause existing functionality to not work as before)_

# Checklist
- [x] Read the [Contributing Guide](https://github.com/streamich/react-use/blob/master/CONTRIBUTING.md)
- [x] Perform a code self-review
- [ ] Comment the code, particularly in hard-to-understand areas
- [x] Add documentation
- [x] Add hook's story at Storybook
- [x] Cover changes with tests
- [x] Ensure the test suite passes (`yarn test`)
- [ ] Provide 100% tests coverage
- [x] Make sure code lints (`yarn lint`). Fix it with `yarn lint:fix` in case of failure.
- [x] Make sure types are fine (`yarn lint:types`).

### CODE REVIEW

File: src/useBodySize.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+const useBodySize = (initialWidth = Infinity, initialHeight = Infinity) => {
+  const [state, setState] = useRafState<{ width: number; height: number }>({
+    width: isClient ? document.body.clientWidth : initialWidth,
+    height: isClient ? document.body.clientHeight : initialHeight,
+  });
+
+  const observer = useMemo(
+    () =>
+      new (window as any).ResizeObserver((entries) => {
```
Comment (ngobach) [inline src/useBodySize.ts:14]: Please correct me if I'm wrong.
`useMemo()` will also be called on SSR requests, won't it? If yes then this line of code might throw on SSR request.

**Comment by ngobach** [comment]
I see your source code is very similar to an existing hook ([**useMeasure**](https://github.com/streamich/react-use/blob/master/src/useMeasure.ts)). It is even possible that your PR can be replaced by using this hook with `ref` pointed to `body`.
Also there is another hook called `useWindowSize` that might be what you wanted.

**Comment by carbongo** [comment]
> I see your source code is very similar to an existing hook ([**useMeasure**](https://github.com/streamich/react-use/blob/master/src/useMeasure.ts)). It is even possible that your PR can be replaced by using this hook with `ref` pointed to `body`.
> Also there is another hook called `useWindowSize` that might be what you wanted.

There is a big difference with `useWindowSize`. `useWindowSize` calculates dimensions of window, but `useBodySize` shows you a size of a document (it's not static and not fully tied to a window size, especially if you have a lot of text, you can see the difference in Storybook). `useBodySize` meant to be called in high-level components, in case you don't want to change the root document for passing dimensions using `useMeasure` through a whole code, you can do so, but... well, considering the fact that I needed those dimensions only in one component, it could've been a bad choice.

**Comment by ngobach** [comment]
Hi @carbongo ,
`useWindowSize` is not what you wanted, so please ignore that one.
About your second point, I think there is a small misunderstanding here, because I didn't mean that you will have to drill dimensions returned from `useBodySize` down to each components. `useMeasure` can be used right in those component with ref set to `document.body`.
If you still see `useBodySize` is an essential hook, I still think you should write it base on `useMeasure` hooks with ref set to `window.body` to avoid duplicated code.

Something like:

```js
function useBodySize() {
  const [setRef, rect] = useMeasure();
  useEffect(() => setRef(document.body));
  return [rect.width, rect.height];
}
```

**Pull request closed without merging**